### PR TITLE
Makeover conf

### PR DESCRIPTION
**Pull request by safl**
With this, then the Python utilities does not rely on having started the interactive cijoe environment.
That is, having invoked:

```bash
cijoe
# [2021-02-11 11:06:28] Welcome to CIJOE ENV_FPATH: /home/safl/.local/share/cijoe/envs/default.sh
[cijoe:default] safl@themachine:~$ 
```

The CIJOE Bash environment is great for doing things interactively, however, it should not be **required** when invoking the test-runner, reporter, etc. although should still utilize it when it is. So, ``cij.conf.Config`` is introduced, producing the installation paths based on the environment, when defined, and based on the result of ``cij_root`` when it is not.

The tools have been updated with the use of the class, and serialization/deserialization added.

### CODE REVIEW

File: modules/cij/runner.py [hunk 1]
```diff
@@ -820,15 +820,17 @@ def main(conf):
...
+            trun_to_junitfile(trun)                 # Persist as jUNIT XML
 
         if tplan.exit(trun) or not tplan.entered:
             trun.status = tplan.status = Status.Fail
         if tplan.status == Status.Unkn and tplan.entered:
             tplan.status = Status.Pass
 
+        trun_to_file(trun)                      # Persist trun
+        trun_to_junitfile(trun)                 # Persist as jUNIT XML
+
```
Comment (micvbang) [inline modules/cij/runner.py:865]: Nice little quality of life upgrade!

File: bin/cij_testcases [hunk 2]
```diff
@@ -11,10 +11,11 @@ import os
 import dataclasses
 from cij.reporter import tcase_parse_descr, dset_to_html
 from cij.runner import TestCase
+import cij.conf
 import cij
 
 
-def construct_dset(evars):
+def construct_dset(conf):
```
Comment (micvbang) [inline bin/cij_testcases:18]: Would be nice to add type annotation

File: bin/cij_testcases [hunk 3]
```diff
@@ -87,17 +83,17 @@ def parse_args():
     args.tmpl_fname = os.path.basename(args.tmpl_fpath)
     args.tmpl_name = os.path.splitext(args.tmpl_fname)[0]
 
-    return args, evars
+    return args, conf
 
 
-def main(args, evars):
+def main(args, conf):
```
Comment (micvbang) [inline bin/cij_testcases:89]: Would be nice to add type annotation for `conf`

File: modules/cij/runner.py [hunk 4]
```diff
@@ -286,7 +288,12 @@ class TestRun(Runnable):
     # pylint: disable=too-many-instance-attributes
 
     ver: str = ""
-    conf: dict = dataclasses.field(default_factory=dict)
+    args: argparse.Namespace = dataclasses.field(
+        default_factory=argparse.Namespace
+    )
+    conf: cij.conf.Config = dataclasses.field(
```
Comment (micvbang) [inline modules/cij/runner.py:294]: These type annotations are great -- we should do the same for some of the other fields, e.g. testcases and testsuites. I'll put that on my TODO

File: modules/cij/runner.py [hunk 5]
```diff
@@ -311,8 +318,18 @@ def from_dict(trun_dict) -> TestRun:
         """
         Deserialize a trun dict into TestRun
         """
+
+        def args_from_dict(adict: dict) -> argparse.Namespace:
+            """Construct argparse.Namespace from dict"""
+
+            namespace = argparse.Namespace()
+            namespace.__dict__.update(adict)
```
Comment (micvbang) [inline modules/cij/runner.py:326]: Neat

File: modules/cij/runner.py [hunk 6]
```diff
@@ -467,10 +484,17 @@ def hooks_setup(trun: TestRun, parent, hnames=None) -> Dict[str, List[Hook]]:
 def trun_to_file(trun: TestRun, fpath=None):
     """Dump the given trun to file"""
 
+    def dict_factory(instance):
+        """huha"""
+
+        return dict(
+            (x[0], vars(x[1])) if x[0] == "args" else x for x in instance
+        )
```
Comment (micvbang) [inline modules/cij/runner.py:492]: I think we can give a more precise docstring than this one :D

File: modules/cij/runner.py [hunk 7]
```diff
@@ -732,7 +763,7 @@ def tplan_setup(trun: TestRun, tplan_fpath) -> TestPlan:
     return tplan
 
 
-def trun_setup(conf) -> TestRun:
+def trun_setup(args, conf) -> TestRun:
```
Comment (micvbang) [inline modules/cij/runner.py:766]: Also type annotation for `conf` here

File: modules/cij/runner.py [hunk 8]
```diff
@@ -790,8 +822,8 @@ def main(conf):
             tsuite.enter(trun)
 
             for tcase in (tc for tc in tsuite.testcases if tsuite.entered):
-                tcase_match = conf.get("TESTCASE_MATCH", None)
-                if tcase_match is None or tcase_match in tcase.name:
+                if args.testcase_match is None or \
+                        args.testcase_match in tcase.name:
```
Comment (micvbang) [inline modules/cij/runner.py:826]: Could be written using parenthesis instead of the line continuation `\`

File: bin/cij_tlint [hunk 9]
```diff
@@ -276,9 +277,9 @@ def main(roots):
 
 if __name__ == "__main__":
 
-    CIJ_ROOT = os.environ.get("CIJ_ROOT")
-    if not (CIJ_ROOT and os.path.exists(CIJ_ROOT)):
-        print("Please set ENV. VAR. CIJ_ROOT, correctly")
+    CONF = cij.conf.from_system()
+    if CONF is None:
+        cij.err("rprt: failed retrieving CIJOE configuration")
```
Comment (micvbang) [inline bin/cij_tlint:282]: Is there a reason this is called `CONF` here? I think it's called `conf` in other places

**Review by micvbang — COMMENTED** [review]
Looks great! This is a very nice overall quality of life improvement in the usage of the tools.
I mostly got minor suggestions and comments on syntax stuff

**Comment by micvbang** [comment]
Also some lovely commit messages btw -- I'll try to learn from them

File: bin/cij_tlint (hunk 9 above)
Comment (safl) [reply bin/cij_tlint:282]: Yeah, it is in global-scope, some PEP says that variables in global-scope must be UPPERCASE.

File: modules/cij/runner.py (hunk 8 above)
Comment (safl) [reply modules/cij/runner.py:826]: Nice! Fixed.

File: modules/cij/runner.py (hunk 7 above)
Comment (safl) [reply modules/cij/runner.py:766]: Good idea. Adding for args as well.

File: modules/cij/runner.py (hunk 6 above)
Comment (safl) [reply modules/cij/runner.py:492]: Hah, true. Fixed.

File: bin/cij_testcases (hunk 2 above)
Comment (safl) [reply bin/cij_testcases:18]: Good idea. Fixed.

File: bin/cij_tlint (hunk 9 above)
Comment (micvbang) [reply bin/cij_tlint:282]: Ah, I didn't realize that -- cool!

**Comment by safl** [comment]
Update based on review.

**Comment by micvbang** [comment]
Great - LGTM

**Pull request merged**